inc/zoo/swar/SWAR.h

```diff
@@ -157,57 +161,152 @@ template<int NBits, typename T = uint64_t> struct SWAR {
+// can bound your inputs at half lane size).
+// This leads to a useful technique (which we use in the robin hood table)
+// where we interleave two related small bit count integers inside of a lane of
+// swar.  More generally, this is useful because it sometimes allows fast
+// operations on side "a" of some lane if side "b" is blitted out, and vice
+// versa.  In the spirit of separation of concerns, we provide a cut-lane-SWAR
+// abstraction here.
+
+template<int NBitsMost, int NBitsLeast, typename T = uint64_t> struct SWARWithSubLanes : SWAR<NBitsMost+NBitsLeast, T> {
+    static constexpr inline auto Available = sizeof(T);
+    static constexpr inline auto LaneBits = NBitsLeast+NBitsMost;
+    static constexpr inline auto NSlots = Available * 8 / LaneBits;
+
+    SWARWithSubLanes() = default;
+    constexpr explicit SWARWithSubLanes(T v) : SWAR<NBitsMost+NBitsLeast, T>(v) {}
+    constexpr explicit operator T() const noexcept { return this->m_v; }
+
+    // M is most significant bits slice, L is least significant bits slice.
+    // 0x....M2L2M1L1 or MN|LN||...||M2|L2||M1|L1
+    using SL = SWARWithSubLanes<NBitsMost, NBitsLeast, T>;
+    using Base = swar::SWAR<LaneBits, T>;
+
+    //constexpr T Ones = makeBitmask<NBits, T>(SWAR<NBits, T>{1}.value());
+    static constexpr inline auto LeastOnes = makeBitmask<LaneBits, T>(Base{1}.value());
+    static constexpr inline auto MostOnes = makeBitmask<LaneBits, T>(Base{1<<NBitsLeast}.value());
+    static constexpr inline auto LeastMask = makeBitmask<LaneBits, T>(Base{~(~0ull<<NBitsLeast)}.value());
+    static constexpr inline auto MostMask = ~LeastMask;
+
+    constexpr auto least() const noexcept {
+        return this->m_v & LeastMask;
+    }
+
+    // Returns only the least significant bits at specified position.
+    constexpr auto least(u32 pos) const noexcept {
+        constexpr auto filter = (T(1) << LaneBits) - 1;
+        const auto keep = (filter << (LaneBits * pos)) & LeastMask;
+        return this->m_v & keep;
+    }
+
+    // Returns only the least significant bits at specified position, 'decoded' to their integer value.
+    constexpr auto leastFlat(u32 pos) const noexcept {
+        return least(pos) >> (LaneBits*pos);
+    }
+
+    constexpr auto most() const noexcept {
+        return this->m_v & MostMask;
+    }
+
+    // Returns only the most significant bits at specified position.
+    constexpr auto most(u32 pos) const noexcept {
+        constexpr auto filter = (T(1) << LaneBits) - 1;
+        const auto keep = (filter << (LaneBits * pos)) & MostMask;
+        return this->m_v & keep;
+    }
+
+    // Returns only the most significant bits at specified position, 'decoded' to their integer value.
+    constexpr auto mostFlat(u32 pos) const noexcept {
+        return most(pos) >> (LaneBits*pos)>> NBitsLeast;
+    }
+
+    // Sets the lsb sublane at |pos| with least significant NBitsLeast of |in|
+    constexpr auto least(T in, u8 pos) const noexcept {
```

Comment:
Please don't use `u8` here.
If you look at the code that deals with indices, they are `int`.
The reason is that `int` is the natural integer size for the architecture, hence it minimizes the chances of penalties of integer size conversions.
Also, it is actually important, however subtle, to keep these *signed* integers, the reason being that integer overflow is "undefined behavior", this means the compiler is allowed to assume the program won't enter undefined behavior and thus won't overflow, then it can enable optimizations it otherwise would not be able to.
I tried to make an illustration of this but in 15 minutes I could not

Reply:
fixed, pushed